src/main/java/group1/mips_simulator/SimController.java

```diff
@@ -69,41 +71,144 @@ protected void restartComputer() {
+        this.computer.mem.write(memAddReg.read(), memBuffReg.read());
+        redraw();
     }
 
+    /**
+     * The same as the storeOnClick function, except this will increment the location
+     * of the Memory Address Register BEFORE writing to memory
+     */
     public void storePlusOnClick(ActionEvent actionEvent) {
+        Register memAddReg = this.computer.cpu.regfile.getMAR();
+        memAddReg.write((short) (memAddReg.read().get() + 1));
+        storeOnClick(actionEvent);
+        redraw();
     }
 
+    /**
+     * Spin up a parallel thread and start running the program on that background thread
+     * This allows the main thread to listen for HALT events
+     */
     public void runOnClick(ActionEvent actionEvent) {
+        // Halt any previous task
+        this.haltOnClick(actionEvent);
+        this.userHalt = false;
+
+        // Run the task in a background thread so the main thread
+        // may listen for interrupt activities (or for other button clicks
+        // the user may make)
+        this.programRunnerThread = new Thread(this::runOnClickTask);
+        // Bind the thread to the main one (to stop both in case of user closing the window)
+        programRunnerThread.setDaemon(true);
+        // Start the program
+        programRunnerThread.start();
     }
 
+    /**
+     * This function is the "task" to be run on the parallel thread.
+     * NOTE: This functions features a (potentially) infinite while(true)
+     * loop. Only a haltOnClick() event, or the program's natural termination
+     * will stop the while(true) loop.
+     */
+    private void runOnClickTask() {
+        boolean computerMayContinue = true; // Track if the program is requesting a halt
+
+        // Run the instruction at the current Program Counter forever until the user
+        // or the computer itself requests a halt.
+        while (!userHalt && computerMayContinue) {
+            System.out.println("Running Instruction");
+            computerMayContinue = this.computer.runCurrentPC();
+        }
+        Platform.runLater(this::redraw);
+    }
+
+    /**
+     * Only run a single instruction, from the current Program Counter.
+     * Redraw the frame afterward.
+     */
     public void stepOnClick(ActionEvent actionEvent) {
+        this.computer.runCurrentPC();
+        this.redraw();
     }
 
+    /**
+     * Halt any existing program (if one is running).
+     */
     public void haltOnClick(ActionEvent actionEvent) {
+        System.out.println("HALTING!!!!!!!!!!!!!!!!!!!!!!!");
+        userHalt = true;
+        if (programRunnerThread != null) {
+            programRunnerThread.interrupt();
+        }
+
+        System.out.println("Halt Redrawing");
+        // redraw the frame after forcing the program to stop early to render the
+        // changes in the Model to the View.
+        redraw();
+        System.out.println("Finished HALTING!!!!!!!!!!!!!!!!!!!!!!!");
```

Comment:
very intense, nice

Reply:
Well, when it runs 5m instructions per second I was afraid that I'd miss this output.